client/src/components/Game/Timer.js

```diff
@@ -0,0 +1,136 @@
+import React, { Component } from 'react';
+import { withFirebase } from '../Firebase';
+import { withRouter } from 'react-router';
+import { compose } from 'recompose';
+
+class Timer extends Component {
+
+  constructor(props) {
+    super(props);
+    this.state = {
+
+    }
+  }
+
+  async componentDidMount() {
+    const { id } = this.props.match.params;
+    const game = this.props.firebase.game(id);
+
+    game.get()
+    .then(docSnapshot => {
+      if(docSnapshot.exists) {
+        this.unsubscribe = game.onSnapshot((snapshot) => {
+          if(snapshot.data().gameStartTime !== null && snapshot.data().currentPlayerIndex === 0) {
+            if(snapshot.data().currentPlayerIndex === 0) {
+              this.setState({
+                startTime: snapshot.data().gameStartTime.seconds,
+                timePerTurnInSeconds: parseInt(snapshot.data().timeLimit, 10) * 60,
+              })
+            } else {
+              this.setState({
+                startTime: Math.floor(new Date().getTime() / 1000),
+                timePerTurnInSeconds: parseInt(snapshot.data().timeLimit, 10) * 60,
+              })
+            }
+            this.setState({
+              timeTurnWillEnd: (this.state.timePerTurnInSeconds + this.state.startTime),
+            })
+            this.setState({
+              days: this.calculateTime('days'),
+              hours: this.calculateTime('hours'),
+              minutes: this.calculateTime('minutes'),
+              seconds: this.calculateTime('seconds')
+            })
+          }
+        })
+      }
+    })
+
+    this.myInterval = setInterval(async () => {
+      const { days, hours, minutes, seconds } = this.state
+      const gameRef = await this.props.firebase.game(id).get();
+
+      if (seconds > 0) {
+        this.setState({
+          seconds: this.calculateTime('seconds')
+        })
+      }
+      if (seconds === 0) {
+        if (minutes === 0 && hours === 0 && days === 0) {
+          this.sendEmail(gameRef, id);
+          game.set({
+            currentPlayerIndex: gameRef.data().currentPlayerIndex + 1,
```

Comment:
Has this been tested? I am concerned the next player's game will give an index out of bounds error when accessing the drawing array.

Reply:
I thought this would probably also break rendering the endgame collage and download in some way, but I don't see those files changed at all

Reply:
why would it break the drawing array? the currentPlayerIndex is only added if the timer runs out or if they press send, never at the same time

Reply:
Has it been tested that it will not break? That when the timer runs out, the next player gets the correct drawing displayed?

Reply:
[This](https://github.com/googleinterns/Pictophone/blob/29f14daf6ba113e934ace88cc96bd139f95d30ad/client/src/components/Game/Canvas.js#L59) is the line of code I think will go OOB. If player index `x` does not send a drawing in time, the drawing array will only have `x` items in it (from players `0..x-1`). When player index `x+1` tries to load the previous img, the canvas will try to display `drawings[x]` which is OOB for an array of length `x`.

Reply:
The endgame code naively assumes player index `i` corresponds to drawing index `i` and [zips them together](https://github.com/googleinterns/Pictophone/blob/29f14daf6ba113e934ace88cc96bd139f95d30ad/client/src/components/Game/Endgame.js#L78-L85). Would this still be the case if a player is skipped?

Reply:
okay so i suggest that we change userIndex to drawings.length() so it always gets the most recent image, what do you think?

Reply:
Note I haven't pulled your code or tested it, this is just speculation.

Reply:
Solution seems to work for displaying previous image on canvas but not for endgame

Reply:
im not sure how we'd fix the names for the photos in the endgame, but my suggestion should fix the OOB error